### PR TITLE
Allow missing key for None fields

### LINKED ISSUE #93: Option cannot handle non-existing fields in decoder
When using Option following code is generated for decoder:
```
value["items"].fold({Either.Left(KeyNotFound("items"))}, { arrow.core.Option.Companion.decoder(ListDecoderInstance(kotlin.String.decoder())).run { decode(it) } }),
```

So when "items" are not in the JSON error is returned despite using Option (which should allow not only values with null, but also non existing values).

### PR DESCRIPTION
**Pull request by AdrianRaFo**
Fixes #93

### CODE REVIEW

**Comment by 47degdev** [comment]
***Hood benchmark comparison:***
:warning: `Decoding` (Threshold: 293.2505758402619)

Benchmark | Value
|--|--|
master_benchmark | 11743.366810921772
helios_benchmark | 11551.261864437243


:red_circle: `Decodingfromraw` (Threshold: 250.0)

Benchmark | Value
|--|--|
master_benchmark | 7531.787365552707
helios_benchmark | 7234.885800282313


:heavy_check_mark: `Parsing` (Threshold: 250.0)

Benchmark | Value
|--|--|
master_benchmark | 22275.27954536463
helios_benchmark | 22317.491938822266

**Comment by AdrianRaFo** [comment]
@nomisRev ready for review, @fedefernandez ready for a second round

File: helios-core/src/main/kotlin/helios/instances/instances.kt [hunk 1]
```diff
@@ -186,4 +187,44 @@ interface TripleDecoderInstance<out A, out B, out C> : Decoder<Triple<A, B, C>>
...
+}
+
+@extension
+interface NullableEncoderInstance<in A> : Encoder<A?> {
+
+  fun encoderA(): Encoder<A>
+
+  override fun A?.encode(): Json {
+    val maybeThis = toOption()
+    return Option.encoder(encoderA()).run { maybeThis.encode() }
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instances.kt]:
```suggestion
    return this?.let { a ->encoderA().let { a.encode() } } ?: JsNull
```

Need multi-line suggestions ><

There is not need for this indirection, it's redundant overhead and when parsing large `Json`s will unnecessary slow down perf.

File: helios-core/src/main/kotlin/helios/instances/instances.kt [hunk 2]
```diff
@@ -186,4 +187,44 @@ interface TripleDecoderInstance<out A, out B, out C> : Decoder<Triple<A, B, C>>
...
+
+}
+
+@extension
+interface NullableDecoderInstance<out A> : Decoder<A?> {
+
+  fun decoderA(): Decoder<A>
+
+  override fun decode(value: Json): Either<DecodingError, A?> =
+    Option.decoder(decoderA()).decode(value).map(Option<A>::orNull)
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instances.kt]:
```suggestion
    if (value.isNull) null.right() else decoderA.decode(value)
```

**Pull request merged**